File: arrayHashing/anagramGroups.js
/**
  * Anagram Groups
Given an array of strings strs, group all anagrams 
together into sublists. You may return the output in any order.

An anagram is a string that contains the exact same characters 
as another string, but the order of the characters can be different.

Example 1:

Input: strs = ["act","pots","tops","cat","stop","hat"]

Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]
Example 2:

Input: strs = ["x"]

Output: [["x"]]
Example 3:

Input: strs = [""]

Output: [[""]]

*/

function groupAnagram(strs) {
    const ans = {};

    for (const str of strs) {
        const count = Array(26).fill(0);

        for (const c of str) {
            count[c.charCodeAt(0) - "a".charCodeAt(0)]++;
        }
        const key = count.join("#");
        if (!ans[key]) {
            ans[key] = [];
        }
        ans[key].push(str);
    }
    return Object.values(ans);
}
strs = ["act", "pots", "tops", "cat", "stop", "hat"];
const answer = groupAnagram(strs);

console.log(answer);

File: arrayHashing/anagramGroups.go
package arrayhashing

import "fmt"

func groupAnagrams(strs []string) [][]string{
	/*
	
    [0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 1 0 0]: ["eat", "tea", "ate"],
    [0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 1 0 0]: ["tan", "nat"],
    [0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]: ["bat"]

	result := [][]string{
	it is a slice that its elements are slices of strings
    {"eat", "tea", "ate"}, // First element, a slice of strings
    {"tan", "nat"},        // Second element, another slice of strings
    {"bat"},               // Third element, another slice of strings
}
    

	*/
    res := make(map[[26]int][]string)

    for _, s := range strs {
        var count [26]int 
        
        for _, c := range s{
            count[c-'a']++
        }
        res[count] = append(res[count], s)
    }

    var result [][]string
    for _, group := range res {
        result = append(result, group)
    }
    return result
    /*
        const x = 3.14 type inference
        var x int
        var x int = 3.14
        x:=3.14 // type inference

        for slices
        arr := []string {"strings", "strings"}
        var arr []string = []string{"hello", "world"}
        var arr = []string{"hello", "world"}

        var arr []string
        arr = []string {"hello", "world"}
    */
}

func main(){
    arr := []string{"act","pots","tops","cat","stop","hat"}
    fmt.Println(groupAnagrams(arr))
}